[WebGPU] A retained MTLSamplerState can result us in exceeding the MTLSamplerState limit

mwyrzykowski opened the pull request:
#### 69057c3c0809b3ed51ea1728810e972aca57297c
<pre>
[WebGPU] A retained MTLSamplerState can result us in exceeding the MTLSamplerState limit
<a href="https://bugs.webkit.org/show_bug.cgi?id=279542">https://bugs.webkit.org/show_bug.cgi?id=279542</a>
<a href="https://rdar.apple.com/135708279">rdar://135708279</a>

Reviewed by Tadeu Zagallo.

Retain samplers until their corresponding command buffer completes.

* Source/WebGPU/WebGPU/BindGroup.h:
(WebGPU::BindGroup::samplers const):
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::addSampler):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setBindGroup):

Canonical link: <a href="https://commits.webkit.org/288127@main">https://commits.webkit.org/288127@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad67d3b066c120e0b90354c7c2f3649bbf2a0721

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81938 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1465 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35895 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/86495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/32970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1500 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9288 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/86495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/32970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85008 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1111 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74562 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/86495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1013 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28739 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31389 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29350 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87926 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9180 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9365 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70381 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/71487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17816 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15588 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14508 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9131 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14667 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8971 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12496 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10779 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Sampler.mm:
```diff
@@ -31,6 +31,11 @@
 #import <wtf/TZoneMallocInlines.h>
 
+// FIXME: remove once FB6018791 is addressed in Foundation
+@interface NSObject (NSPointerArray_15396578)
```
Mike Ash suggested NSPointerArray instead of overriding dealloc as originally done in https://commits.webkit.org/283645@main

However, there is an outstanding bug tracked by FB6018791 / radar://15396578 so with some reflection it can be concluded that `_markNeedsCompaction` needs to be called prior to calling `compact` for `compact` to work as intended.

kkinnunen-apple commented on LayoutTests/fast/webgpu/nocrash/fuzz-279542-expected.txt:
```diff
@@ -0,0 +1 @@
+Test passes if it does not crash.
```
I don't think you can check in a 33k line obfuscated testcase..
Your problem is that the page creates too many samplers?
Could you create a simple page 30 line page that creates too many samplers?

mwyrzykowski replied:
Yes, a reduction would certainly be useful. I will add one.

It is somewhat beneficial to commit the 33k line fuzzer case, since it does not run in EWS (only nightly) to ensure any crashes caught by the fuzzer case do not regress, but a reduction for this issue which runs as part of EWS would be good to add.

kkinnunen-apple commented on Source/WebGPU/WebGPU/Sampler.mm:
```diff
@@ -272,41 +289,62 @@ static uint64_t floatToUint64(float f)
-    if (!cachedSamplerStates) {
-        cachedSamplerStates = [NSMutableDictionary dictionary];
-        lastAccessedKeys = [NSMutableOrderedSet orderedSet];
+    {
```
The samplers seem to be per-device, and that's described by the docs too. The code has samplers in global array. Does not look correct?

mwyrzykowski replied:
Per https://developer.apple.com/documentation/metal/buffers/improving_cpu_performance_by_using_argument_buffers argument buffer samplers are limited 'per app' which means per-process

kkinnunen-apple commented on Source/WebGPU/WebGPU/Sampler.mm:
```diff
@@ -272,41 +289,62 @@ static uint64_t floatToUint64(float f)
         return nil;
-    if (cachedSamplerStates.count >= device.maxArgumentBufferSamplerCount) {
+
+    if (aliveCount() >= device.maxArgumentBufferSamplerCount) {
```
The device is used by CA and WebGL too. I don't think you can use the max amount of samplers, other code breaks.

mwyrzykowski replied:
CA and WebGL don't set `supportArgumentBuffers=YES` https://developer.apple.com/documentation/metal/mtlsamplerdescriptor/2915782-supportargumentbuffers?language=objc

Indeed if they started doing so, we would need to adopt a new strategy, like swizzling `+[_MTLSampler load]` or something

kkinnunen-apple commented on Source/WebGPU/WebGPU/Sampler.mm:
```diff
@@ -272,41 +289,62 @@ static uint64_t floatToUint64(float f)
-    if (cachedSamplerStates.count >= device.maxArgumentBufferSamplerCount) {
+
+    if (aliveCount() >= device.maxArgumentBufferSamplerCount) {
+        Locker locker { samplerStateLock };
```
The locking here is dubious. You do:
- lock
- free
- unlock
- allocate

If the locking is truely needed, then somebody can race free and alloc.

Consider algorithm of something like:

```
lock
do (
   alloc()
   if succeed break;
   free();
) while (can free)
unlock
```

If contention is an issue, try moving the lock inside.

kkinnunen-apple commented on Source/WebGPU/WebGPU/Sampler.mm:
```diff
@@ -272,41 +289,62 @@ static uint64_t floatToUint64(float f)
+        if (!cachedSamplerStates) {
+            cachedSamplerStates = [NSMutableDictionary dictionary];
+            lastAccessedKeys = [NSMutableOrderedSet orderedSet];
+            weakSamplerStates = [NSPointerArray weakObjectsPointerArray];
```
Your commit messaqge does not really describe what problem you are trying to solve with weakSamplerStates.

It does not help that the testcase is 33k obfuscated blob.

Is the problem that even though WebGPU has freed its internal cache entries, some other part of the system might still hold the refs to the samplers in the freed cache entries, and thus freeing did nothing? And you solve this by tracking the weak ptrs?

mwyrzykowski replied:
yes, effectively Metal command buffers will still retain references to the MTLSamplerStates in use by the command buffers, so a long running Metal command buffer can cause this race to occur with the existing sampler state cache.

kkinnunen-apple commented on Source/WebGPU/WebGPU/Sampler.mm:
```diff
@@ -272,41 +289,62 @@ static uint64_t floatToUint64(float f)
+
 id<MTLSamplerState> Sampler::samplerState() const
 {
     if (!m_samplerIdentifier)
```
For clarity, Here you could do also:

```
id<MTLSamplerState> samplerState = m_cachedSamplerState;
if (samplerState)
   return samplerState;
```

kkinnunen-apple commented on Source/WebGPU/WebGPU/Sampler.mm:
```diff
@@ -272,41 +289,62 @@ static uint64_t floatToUint64(float f)
+    if (aliveCount() >= device.maxArgumentBufferSamplerCount)
+        return nil;
+
+    id<MTLSamplerState> samplerState = [device newSamplerStateWithDescriptor:createMetalDescriptorFromDescriptor(m_descriptor)];
```
If you get null from the system it means there is too many samplers, either yours or somebody else's.
What you can do is free your own -- you cannot free anybody elses'.
So you should error out once you're freed all your own and it still doesn't give out samplers.


I think the crucial thing to understand is that `aliveCount()` counts the "sampler objects WebGPU has created".
If you are going to create max amount of objects as in this patch, that number really doesn't matter one bit. 
The simpler algorithm is that you just create the objects until you cannot. Who cares if aliveCount() returns 5 and the allocation fails? Who cares if the aliveCount() returns  2048 and the allocation fails?

It matters only if you are going to limit to something sane. E.g "Out of 2048 available unique sampler objects, WebGPU can create 20 and then error out".

I think the algorithm needs to be thought out again.

My suggestion:
* WebGPU should limit its use to something sane
* The limit should be *total in-use objects*. E.g. use the weak pointer machinery similar to this patch.
* Allocation should be a loop that allocates and if allocation fails, it frees. The loop continues until there's nothing to free.

mwyrzykowski replied:
The concern with allowing `newSamplerStateWithDescriptor:` to return nil is the metal validation layer release asserts when the limit is exceeded. I would need to check the driver source what occurs when we exceed the global argument buffer process sampler limit with validation disabled to see if we can rely on this function returning nil.

kkinnunen-apple replied:
Ok, I see. It seems we need to track the created MTLSamplerStates with the weak ptrs.
Well, could we still fix the allocation as a loop and try to fix the mutexing to avoid the obvious races.

mwyrzykowski committed:
[WebGPU] A retained MTLSamplerState can result us in exceeding the MTLSamplerState limit
https://bugs.webkit.org/show_bug.cgi?id=279542
rdar://135708279

Reviewed by Tadeu Zagallo.

Retain samplers until their corresponding command buffer completes.

* Source/WebGPU/WebGPU/BindGroup.h:
(WebGPU::BindGroup::samplers const):
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::addSampler):
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setBindGroup):

Canonical link: https://commits.webkit.org/288127@main
